lib/puppet/configurer.rb

```diff
@@ -49,6 +51,14 @@ def initialize(factory = Puppet::Configurer::DownloaderFactory.new)
     @splayed = false
     @environment = Puppet[:environment]
     @transaction_uuid = SecureRandom.uuid
+    @static_catalog = true
+
+    # Get checksum types known to the agent, preferring the default specified
+    # in the digest_algorithm setting.
+    @checksum_type = valid_checksum_types
```

Comment:
I guess at that point I'd lean towards a single checksum type. If the agent default isn't supported by the master, they can just set `digest_algorithm` on their agents.